### PR TITLE
Blazor client-side validation

### LINKED ISSUE #5524: Interactive Validation model for the Blazor UI
Can we automate or simplify the validation stuff, with localization?

* Should support data annotation validations
* Should support IValidatableObject
* Should support FluentValidation (with low priority and we can create a separate issue for that)

Initially, we can implement the validation for the user create and edit modals of the identity module and then we discuss on it (before deciding to implement for other modules).

### LINKED ISSUE #5915: Create additional before and after methods for AbpCrudPageBase operations

### PR DESCRIPTION
**Pull request by stsrki**
Do not merge! This is still a work in progress.

- Added client-side validation to roles and users
- Localization of error messages
- Localization of arguments for the error message
- Publish new Blazorise and update APB project references

### CODE REVIEW

**Comment by stsrki** [comment]
@hikalkan Please review this PR once you have time. I think it should be ready now.

We still need to do FluentValidation but that can be done later.

**Comment by stsrki** [comment]
@hikalkan Please review when you're ready?

Validation is now working as expected. Also, I have merged and fixed conflicts with dev branch.

File: framework/src/Volo.Abp.BlazoriseUI/AbpCrudPageBase.cs [hunk 1]
```diff
@@ -279,10 +281,31 @@ protected virtual async Task OnDataGridReadAsync(DataGridReadDataEventArgs<TList
...
 
             NewEntity = new TCreateViewModel();
+
+            // Mapper will not notify Blazor that binded values are changed
+            // so we need to notify it manually by calling StateHasChanged
+            await InvokeAsync(() => StateHasChanged());
+
             CreateModal.Show();
+
+            await OnOpeningCreateModalAsync();
```
Comment (hikalkan) [inline framework/src/Volo.Abp.BlazoriseUI/AbpCrudPageBase.cs]:
```suggestion
            await OnOpenedCreateModalAsync();
```

But I think there is no point to add these methods now. I can override the `OpenCreateModalAsync` and make anything before and after it. Just remove them please.

File: framework/src/Volo.Abp.BlazoriseUI/AbpCrudPageBase.cs [hunk 2]
```diff
@@ -293,12 +316,32 @@ protected virtual Task CloseCreateModalAsync()
...
             EditingEntity = MapToEditingEntity(entityDto);
+
+            await InvokeAsync(() => StateHasChanged());
+
             EditModal.Show();
+
+            await OnOpenedEditModalAsync(id);
+        }
+
+        protected virtual Task OnOpeningEditModalAsync(TKey id)
```
Comment (hikalkan) [inline framework/src/Volo.Abp.BlazoriseUI/AbpCrudPageBase.cs:328]: Same.. we can remove these methods.

File: framework/src/Volo.Abp.Core/System/AbpObjectExtensions.cs [hunk 3]
```diff
@@ -105,5 +106,62 @@ public static T If<T>(this T obj, bool condition, Action<T> action)
...
         }
+
+        /// <summary>
+        /// Resets all fields of a given object to their default values.
+        /// </summary>
+        /// <typeparam name="T">Type of the object.</typeparam>
+        /// <param name="obj">An object</param>
+        /// <param name="customReset">Alternative custom method to reset the values.</param>
+        /// <returns>Returns the original object.</returns>
+        public static T ResetState<T>(this T obj, Action<T> customReset = null)
```
Comment (hikalkan) [inline framework/src/Volo.Abp.Core/System/AbpObjectExtensions.cs]: If we don't use this method anymore, please delete it.

File: framework/src/Volo.Abp.Core/System/AbpObjectExtensions.cs [hunk 4]
```diff
@@ -105,5 +106,62 @@ public static T If<T>(this T obj, bool condition, Action<T> action)
...
+        }
+
+        /// <summary>
+        /// Applies all fields of a given object from it's counterpart.
+        /// </summary>
+        /// <typeparam name="T">Type of the object.</typeparam>
+        /// <param name="obj">An object.</param>
+        /// <param name="other">An object from which we copy the values.</param>
+        /// <returns>Returns the original object.</returns>
+        public static T ApplyState<T>(this T obj, T other)
```
Comment (hikalkan) [inline framework/src/Volo.Abp.Core/System/AbpObjectExtensions.cs]: If we don't use this method anymore, please delete it.

File: modules/identity/src/Volo.Abp.Identity.Blazor/AbpIdentityBlazorMessageLocalizerExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Blazorise;
+using Microsoft.Extensions.Localization;
+
+namespace Volo.Abp.Identity.Blazor
+{
+    public static class AbpIdentityBlazorMessageLocalizerExtensions
+    {
+        public static string Localize<T>(this IStringLocalizer<T> stringLocalizer, string format, IEnumerable<string> arguments)
```
Comment (hikalkan) [inline modules/identity/src/Volo.Abp.Identity.Blazor/AbpIdentityBlazorMessageLocalizerExtensions.cs]: This method name is too general and it is especially dengerous when it is an extension method.
I suggest to not make this an extension method and change the class name as `AbpIdentityBlazorMessageLocalizerHelper` or something like that. If it is only for displayname style localization, the method name can indicate that.

File: modules/identity/src/Volo.Abp.Identity.Blazor/AbpIdentityBlazorMessageLocalizerExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            try
+            {
+                return arguments?.Count() > 0
+                    ? string.Format(stringLocalizer[format], LocalizeMessageArguments(stringLocalizer, arguments)?.ToArray())
+                    : stringLocalizer[format];
```
Comment (hikalkan) [inline modules/identity/src/Volo.Abp.Identity.Blazor/AbpIdentityBlazorMessageLocalizerExtensions.cs]: `stringLocalizer[...]` has already an overload like `stringLocalizer[format, arguments]`. Why don't you use it?

Also, `IEnumerable<string> arguments` parameter can be made optional if it allows nulls, like `[CanBeNull] IEnumerable<string> arguments = null`.

Finally, as I understand, `format` can be a simple string that has no format arguments. In this case, a `message` parameter name would be better.

File: modules/identity/src/Volo.Abp.Identity.Blazor/Pages/Identity/RoleManagement.razor [hunk 7]
```diff
@@ -1,17 +1,19 @@
 ﻿@page "/identity/roles"
-@attribute [Authorize(IdentityPermissions.Roles.Default)]
+@attribute [Authorize( IdentityPermissions.Roles.Default )]
```
Comment (hikalkan) [inline modules/identity/src/Volo.Abp.Identity.Blazor/Pages/Identity/RoleManagement.razor:2]: Don't use extra spaces here, please change the editor formatting config :)

File: modules/identity/src/Volo.Abp.Identity.Blazor/Pages/Identity/RoleManagement.razor [hunk 8]
```diff
@@ -28,7 +30,7 @@
           ShowPager="true"
           PageSize="PageSize">
     <DataGridColumns>
-        @if (ShouldShowEntityActions)
+        @if ( ShouldShowEntityActions )
```
Comment (hikalkan) [inline modules/identity/src/Volo.Abp.Identity.Blazor/Pages/Identity/RoleManagement.razor:34]: Again, we don't use extra spaces. Please change the editor settings. Otherwise, we see a lot of unnecessary changes whenever you touch a code file.

**Review by hikalkan — CHANGES_REQUESTED** [review]

File: modules/identity/src/Volo.Abp.Identity.Blazor/Pages/Identity/RoleManagement.razor (hunk 8 above)
Comment (stsrki) [reply modules/identity/src/Volo.Abp.Identity.Blazor/Pages/Identity/RoleManagement.razor:34]: Fixed

File: framework/src/Volo.Abp.BlazoriseUI/AbpCrudPageBase.cs (hunk 1 above)
Comment (stsrki) [reply framework/src/Volo.Abp.BlazoriseUI/AbpCrudPageBase.cs]: I removed OnOpened and leave only OnOpening methods since we're using those

File: framework/src/Volo.Abp.Core/System/AbpObjectExtensions.cs (hunk 3 above)
Comment (stsrki) [reply framework/src/Volo.Abp.Core/System/AbpObjectExtensions.cs]: Removed

File: framework/src/Volo.Abp.BlazoriseUI/AbpCrudPageBase.cs (hunk 2 above)
Comment (stsrki) [reply framework/src/Volo.Abp.BlazoriseUI/AbpCrudPageBase.cs:328]: I would leave OnOpeningEditModalAsync because we use it

File: framework/src/Volo.Abp.Core/System/AbpObjectExtensions.cs (hunk 4 above)
Comment (stsrki) [reply framework/src/Volo.Abp.Core/System/AbpObjectExtensions.cs]: Removed

File: modules/identity/src/Volo.Abp.Identity.Blazor/AbpIdentityBlazorMessageLocalizerExtensions.cs (hunk 5 above)
Comment (stsrki) [reply modules/identity/src/Volo.Abp.Identity.Blazor/AbpIdentityBlazorMessageLocalizerExtensions.cs]: The reason why I wanted it as an extension method id because it's easier to write `L.Localize` then something like `BlazorMessageLocalizer.Localize(stringLocalizer, message, args)`

File: modules/identity/src/Volo.Abp.Identity.Blazor/AbpIdentityBlazorMessageLocalizerExtensions.cs (hunk 6 above)
Comment (stsrki) [reply modules/identity/src/Volo.Abp.Identity.Blazor/AbpIdentityBlazorMessageLocalizerExtensions.cs]: I have now used `stringLocalizer[format, arguments]`. And the reason for all of this code is because I need to localize arguments along with messages because they have "fallback" logic.

```
return stringLocalizer[$"DisplayName:{argument}"]
  ?? stringLocalizer[argument]
  ?? argument;
```

File: modules/identity/src/Volo.Abp.Identity.Blazor/AbpIdentityBlazorMessageLocalizerExtensions.cs (hunk 5 above)
Comment (stsrki) [reply modules/identity/src/Volo.Abp.Identity.Blazor/AbpIdentityBlazorMessageLocalizerExtensions.cs]: Created AbpIdentityBlazorMessageLocalizerHelper instead of extension methods
